### PR TITLE
Resolving ILLink warnings on Microsoft.Extensions.Options.ConfigurationManager

### PR DESCRIPTION
**Pull request by joperezr**
Resolving ILLink warnings on Microsoft.Extensions.Options.ConfigurationManager

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Logging.Configuration/ref/Microsoft.Extensions.Logging.Configuration.cs [hunk 1]
```diff
@@ -23,7 +23,8 @@ public partial interface ILoggerProviderConfiguration<T>
     }
     public static partial class LoggerProviderOptions
     {
-        public static void RegisterProviderOptions<TOptions, TProvider>(Microsoft.Extensions.DependencyInjection.IServiceCollection services) where TOptions : class { }
+        [System.Diagnostics.CodeAnalysis.RequiresUnreferencedCode("The type of TOptions cannot be statically analyzed so its members may be trimmed.")]
```
Comment (vitek-karas) [inline src/libraries/Microsoft.Extensions.Logging.Configuration/ref/Microsoft.Extensions.Logging.Configuration.cs:26]: I think the message is somewhat misleading. Generic parameters are typically statically analyzable, that's not the problem here. The problem is that we don't have annotation which would recursively preserve enough things on TOptions to work all the time.

Maybe something more generic:
"Trimming may remove some parts of the TOptions and it's dependent types"

Can't come up with something good right now, hopefully somebody will have a good idea.

File: src/libraries/Microsoft.Extensions.Options.ConfigurationExtensions/src/ConfigureFromConfigurationOptions.cs [hunk 2]
```diff
@@ -12,20 +14,26 @@ namespace Microsoft.Extensions.Options
...
     /// <typeparam name="TOptions">The type of options to bind.</typeparam>
-    public class ConfigureFromConfigurationOptions<TOptions> : ConfigureOptions<TOptions>
+    public class ConfigureFromConfigurationOptions<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] TOptions> : ConfigureOptions<TOptions>
         where TOptions : class
     {
         /// <summary>
         /// Constructor that takes the <see cref="IConfiguration"/> instance to bind against.
         /// </summary>
         /// <param name="config">The <see cref="IConfiguration"/> instance.</param>
+        //Even though TOptions is annotated, we need to annotate as RUC as we can't guarantee that options that are passed in are a nested type.
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Options.ConfigurationExtensions/src/ConfigureFromConfigurationOptions.cs:24]: This comment is not correct.

"we can't guarantee that options that are passed in are a nested type."

should say

"we can't guarantee properties on referenced types are preserved."

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerExtensions.cs [hunk 3]
```diff
@@ -16,10 +16,15 @@ namespace Microsoft.Extensions.Logging
     [UnsupportedOSPlatform("browser")]
     public static class ConsoleLoggerExtensions
     {
+        internal const string TrimmingRequiresUnreferencedCodeMessage = "The type of TOptions cannot be statically analyzed so its members may be trimmed.";
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerExtensions.cs:19]: (nit) `private`

File: src/libraries/Microsoft.Extensions.Logging.Configuration/ref/Microsoft.Extensions.Logging.Configuration.cs (hunk 1 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Logging.Configuration/ref/Microsoft.Extensions.Logging.Configuration.cs:26]: I'd also add a suggestion of what to do in order to make it trim-compatible. Something like "Ensure the necessary constructors and properties are preserved"

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerExtensions.cs (hunk 3 above)
Comment (joperezr) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerExtensions.cs:19]: It actually does have to be internal as I use it in line 171 which is in a different class.

Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerExtensions.cs]: Oh, weird. Good call.

File: src/libraries/Microsoft.Extensions.Logging.Configuration/ref/Microsoft.Extensions.Logging.Configuration.cs [hunk 4]
```diff
@@ -23,7 +23,8 @@ public partial interface ILoggerProviderConfiguration<T>
     }
     public static partial class LoggerProviderOptions
     {
-        public static void RegisterProviderOptions<TOptions, TProvider>(Microsoft.Extensions.DependencyInjection.IServiceCollection services) where TOptions : class { }
+        [System.Diagnostics.CodeAnalysis.RequiresUnreferencedCode("Some parts of TOptions and its dependent types may be trimmed. Ensure the necessary constructors and properties are preserved.")]
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Logging.Configuration/ref/Microsoft.Extensions.Logging.Configuration.cs:26]: How about 

```suggestion
        [System.Diagnostics.CodeAnalysis.RequiresUnreferencedCode("TOptions's dependent types may have their members trimmed. Ensure all necessary members are preserved.")]
```

?

**Review by eerhardt — APPROVED** [review]
Just some nits on the message wording. Other than that, LGTM

File: src/libraries/Microsoft.Extensions.Logging.Configuration/src/LoggerProviderConfigurationExtensions.cs [hunk 5]
```diff
@@ -12,13 +13,16 @@ namespace Microsoft.Extensions.Logging.Configuration
     /// </summary>
     public static class LoggerProviderOptions
     {
+        internal const string TrimmingRequiresUnreferencedCodeMessage = "Some parts of TOptions and its dependent types may be trimmed. Ensure the necessary constructors and properties are preserved.";
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Logging.Configuration/src/LoggerProviderConfigurationExtensions.cs:16]: I think the `src` file needs updating to match the `ref`.

We should also update the other libraries' messages as well to read the same.

Comment (joperezr) [reply src/libraries/Microsoft.Extensions.Logging.Configuration/src/LoggerProviderConfigurationExtensions.cs:16]: good catch, will fix.

**Review by eerhardt — APPROVED** [review]

**Pull request merged**